hapi-fhir-validation/src/main/java/org/hl7/fhir/common/hapi/validation/validator/VersionSpecificWorkerContextWrapper.java

```diff
@@ -806,20 +790,62 @@ public void validateCodeBatchByRef(
 
 	@Nonnull
 	private ValidationResult doValidation(
-			IBaseResource theValueSet,
+			@Nullable ValueSet theValueSet,
 			ConceptValidationOptions theValidationOptions,
 			String theSystem,
 			String theCode,
 			String theDisplay) {
+
+		IBaseResource convertedVs = getOrConvertValueSet(theValueSet);
+
 		IValidationSupport.CodeValidationResult result;
-		if (theValueSet != null) {
-			result = validateCodeInValueSet(theValueSet, theValidationOptions, theSystem, theCode, theDisplay);
+		if (convertedVs != null) {
+			result = validateCodeInValueSet(convertedVs, theValidationOptions, theSystem, theCode, theDisplay);
 		} else {
 			result = validateCodeInCodeSystem(theValidationOptions, theSystem, theCode, theDisplay);
 		}
 		return convertValidationResult(theSystem, result);
 	}
 
+	private IBaseResource getOrConvertValueSet(ValueSet theValueSet) {
+		if (theValueSet == null) {
+			return null;
+		}
+
+		IBaseResource convertedValueSet = (IBaseResource) theValueSet.getUserData(FROM_CANONICAL_USERDATA_KEY);
+		if (convertedValueSet != null) {
+			ourLog.debug(
+					"Using user data - Key: {}, ValueSet: {}",
+					FROM_CANONICAL_USERDATA_KEY,
+					toResourceId(convertedValueSet));
```

Comment:
In this code, the `toResourceId(..)` string will be built regardless of whether we're logging at debug level or not. Recommend using the following form instead:

```suggestion
		ourLog
			.atDebug()
			.setMessage("Using user data - Key: {}, ValueSet: {}")
			.addArgument(FROM_CANONICAL_USERDATA_KEY)
			.addArgument(()->toResourceId(convertedValueSet))
			.log();
```

Reply:
Ahh good to know! Fixed here:

https://github.com/hapifhir/hapi-fhir/pull/6870/commits/746aaf4c35a1ca755d2d3f2dd0c93d87d264eb5c